crates/bevy_math/src/curve.rs

```diff
@@ -0,0 +1,905 @@
+            return Err(InfiniteIntervalError);
+        }
+        let scale = other.length() / self.length();
+        Ok(move |x| (x - self.start) * scale + other.start)
+    }
+
+    /// Get an iterator over `points` equally-spaced points from this interval in increasing order.
+    /// Returns `None` if `points` is less than 2; the spaced points always include the endpoints.
+    pub fn spaced_points(self, points: usize) -> Option<impl Iterator<Item = f32>> {
+        if points < 2 {
+            return None;
+        }
+        let step = self.length() / (points - 1) as f32;
+        Some((0..points).map(move |x| self.start + x as f32 * step))
+    }
+}
+
+impl TryFrom<RangeInclusive<f32>> for Interval {
+    type Error = InvalidIntervalError;
+    fn try_from(range: RangeInclusive<f32>) -> Result<Self, Self::Error> {
+        Interval::new(*range.start(), *range.end())
+    }
+}
+
+/// A trait for types whose values can be intermediately interpolated between two given values
+/// with an auxiliary parameter.
+pub trait Interpolable: Clone {
+    /// Interpolate between this value and the `other` given value using the parameter `t`.
+    /// Note that the parameter `t` is not necessarily clamped to lie between `0` and `1`.
+    fn interpolate(&self, other: &Self, t: f32) -> Self;
+}
+
+impl<S, T> Interpolable for (S, T)
+where
+    S: Interpolable,
+    T: Interpolable,
+{
+    fn interpolate(&self, other: &Self, t: f32) -> Self {
+        (
+            self.0.interpolate(&other.0, t),
+            self.1.interpolate(&other.1, t),
+        )
+    }
+}
+
+impl<T> Interpolable for T
+where
+    T: VectorSpace,
+{
+    fn interpolate(&self, other: &Self, t: f32) -> Self {
+        self.lerp(*other, t)
+    }
+}
+
+impl Interpolable for Quat {
+    fn interpolate(&self, other: &Self, t: f32) -> Self {
+        self.slerp(*other, t)
+    }
+}
+
+/// An error indicating that a resampling operation could not be performed because of
+/// malformed inputs.
+#[derive(Debug)] // TODO: Make this an actual Error.
+pub enum ResamplingError {
+    /// This resampling operation was not provided with enough samples to have well-formed output.
+    NotEnoughSamples(usize),
+    /// This resampling operation failed because of an unbounded interval.
+    InfiniteInterval(InfiniteIntervalError),
+}
+
+/// A trait for a type that can represent values of type `T` parametrized over a fixed interval.
+/// Typical examples of this are actual geometric curves where `T: VectorSpace`, but other kinds
+/// of interpolable data can be represented instead (or in addition).
+pub trait Curve<T>
+where
+    T: Interpolable,
+{
+    /// The interval over which this curve is parametrized.
+    fn domain(&self) -> Interval;
+
+    /// Sample a point on this curve at the parameter value `t`, extracting the associated value.
+    fn sample(&self, t: f32) -> T;
+
+    /// Sample a point on this curve at the parameter value `t`, returning `None` if the point is
```

Comment:
Yeah, that's a good question. Generally speaking, the idea is that sampling behavior outside of the domain interval is implementation-specific. We might want to consider renaming some of the methods to discourage using unchecked sampling in user-space, but it's pretty important to have access to it in the API, since doing otherwise introduces required overhead.